### PR TITLE
Make `F.phase_vocoder` and `T.TimeStretch` handle complex dtype

### PR DESCRIPTION
**Pull request by mthrok**
This PR supersedes #758 and includes further updates.
See #1337 for the overview.

1. `F.phase_vocoder` accepts Tensor with complex dtype.
    * The implementation path has been updated from #758 so that they share the same code path by internally converting the input Tensor to complex dtype and performing all the operation on top of it.
    * Adopted `torch.polar` for simpler Tensor generation from magnitude and angle.
2. Updated tests
    * librosa compatibility test for complex dtype and pseudo complex dtype
        * Extracted the output shape check test and moved it to functional so that it will be tested on all the combination of `{CPU | CUDA} x {complex64 | complex128}`
    * TorchScript compatibility test for `F.phase_vocoder` and `T.TimeStretch`.
    * batch consistency test for `T.TimeStretch`.

## Benchmark of `phase_vocoder`

<details><summary>env</summary>

```
Collecting environment information...
PyTorch version: 1.9.0.dev20210329
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.18.4

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 450.80.02
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] pytorch-sphinx-theme==0.0.24
[pip3] torch==1.9.0.dev20210329
[pip3] torchaudio==0.9.0a0+550dc90
[pip3] torchtext==0.9.0a0+c072ba6
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.1.243             h6bb024c_0
[conda] magma-cuda101             2.5.2                         1    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-include               2020.4             h726a3e6_304    conda-forge
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.3.0            py38h54f3939_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.2           py38h54aff64_0
[conda] numpy-base                1.19.2           py38hfa32c7d_0
[conda] pytorch                   1.9.0.dev20210329 py3.8_cuda10.1_cudnn7.6.3_0    pytorch-nightly
[conda] pytorch-sphinx-theme      0.0.24                    dev_0    <develop>
[conda] torch                     1.7.1                    pypi_0    pypi
[conda] torchaudio                0.9.0a0+550dc90           dev_0    <develop>
[conda] torchtext                 0.9.0a0+c072ba6           dev_0    <develop>
```

</details>

### CPU
<details><summary>code</summary>

```bash
#!/usr/bin/env bash

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -s """
import torch
import torchaudio;

num_channels = 2
num_freq = 1025
num_frames = 400

rate = 0.5
hop_length = 256

torch.manual_seed(0);
spec = torch.randn(num_channels, num_freq, num_frames, 2, dtype=torch.float32);
phase_advance = torch.linspace(0, 3.14 * hop_length, num_freq, dtype=torch.float32)[..., None];
""" """
torchaudio.functional.phase_vocoder(spec, rate, phase_advance)
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -s """
import torch
import torchaudio;

num_channels = 2
num_freq = 1025
num_frames = 400

rate = 0.5
hop_length = 256

torch.manual_seed(0);
spec = torch.view_as_complex(torch.randn(num_channels, num_freq, num_frames, 2, dtype=torch.float32));
phase_advance = torch.linspace(0, 3.14 * hop_length, num_freq, dtype=torch.float32)[..., None];
""" """
torchaudio.functional.phase_vocoder(spec, rate, phase_advance)
"""
```
</details>

| Code                     | Pseudo Complex | Native Complex |
| ------------------------ | -------------: | -------------: |
| 512c2fa (before this PR) |            758 |            N/A |
| With this PR applied     |             97 |             95 |

unit: `msec`

### CUDA
<details><summary>code</summary>

```bash
#!/usr/bin/env bash

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -s """
import torch
import torchaudio;

num_channels = 2
num_freq = 1025
num_frames = 400

rate = 0.5
hop_length = 256

torch.manual_seed(0);
spec = torch.randn(num_channels, num_freq, num_frames, 2, dtype=torch.float32, device='cuda');
phase_advance = torch.linspace(0, 3.14 * hop_length, num_freq, dtype=torch.float32, device='cuda')[..., None];
""" """
torchaudio.functional.phase_vocoder(spec, rate, phase_advance)
"""

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -s """
import torch
import torchaudio;

num_channels = 2
num_freq = 1025
num_frames = 400

rate = 0.5
hop_length = 256

torch.manual_seed(0);
spec = torch.view_as_complex(torch.randn(num_channels, num_freq, num_frames, 2, dtype=torch.float32, device='cuda'));
phase_advance = torch.linspace(0, 3.14 * hop_length, num_freq, dtype=torch.float32, device='cuda')[..., None];
""" """
torchaudio.functional.phase_vocoder(spec, rate, phase_advance)
"""
```
</details>

| Code                     | Pseudo Complex | Native Complex |
| ------------------------ | -------------: | -------------: |
| 512c2fa (before this PR) |           1.85 |            N/A |
| With this PR applied     |           1.22 |           1.21 |

unit: `msec`

### CODE REVIEW

File: torchaudio/functional/functional.py [hunk 1]
```diff
@@ -626,9 +646,11 @@ def phase_vocoder(
     complex_specgrams_stretch = torch.stack([real_stretch, imag_stretch], dim=-1)
```
Comment (mthrok) [inline torchaudio/functional/functional.py:646]: @anjali411 Here, we are constructing complex tensor from magnitude and angle. Is there a more elegant way to do this?

Comment (anjali411) [reply torchaudio/functional/functional.py:646]: yeah `torch.polar(abs, angle)`

File: test/torchaudio_unittest/transforms/batch_consistency_test.py [hunk 2]
```diff
@@ -130,27 +130,28 @@ def test_batch_mfcc(self):
         computed = torchaudio.transforms.MFCC()(waveform.repeat(3, 1, 1))
         self.assertEqual(computed, expected, atol=1e-4, rtol=1e-5)
 
-    def test_batch_TimeStretch(self):
+    def _assert_batch_TimeStretch(self, complex):
```
Comment (mthrok) [inline test/torchaudio_unittest/transforms/batch_consistency_test.py:133]: TODO: simplify this test

File: test/torchaudio_unittest/transforms/torchscript_consistency_impl.py [hunk 3]
```diff
@@ -58,16 +59,6 @@ def test_MuLawDecoding(self):
         tensor = torch.rand((1, 10))
         self._assert_consistency(T.MuLawDecoding(), tensor)
 
-    def test_TimeStretch(self):
```
Comment (anjali411) [inline test/torchaudio_unittest/transforms/torchscript_consistency_impl.py:61]: I think we should still keep the old test to validate that we are not introducing bc breaking changes

File: torchaudio/transforms.py [hunk 4]
```diff
@@ -736,7 +736,10 @@ def forward(self, complex_specgrams: Tensor, overriding_rate: Optional[float] =
         Returns:
             Tensor: Stretched complex spectrogram of dimension (..., freq, ceil(time/rate), complex=2).
         """
-        assert complex_specgrams.size(-1) == 2, "complex_specgrams should be a complex tensor, shape (..., complex=2)"
+        if not complex_specgrams.is_complex() and complex_specgrams.size(-1) != 2:
+            raise ValueError(
+                "complex_specgrams must be either complex dtype or "
```
Comment (anjali411) [inline torchaudio/transforms.py:741]: nit - `must be either native complex tensors or real valued tensors with shape (..., 2)`

File: test/torchaudio_unittest/transforms/torchscript_consistency_impl.py (hunk 3 above)
Comment (anjali411) [reply test/torchaudio_unittest/transforms/torchscript_consistency_impl.py:61]: oh nvm just saw that you unified the tests for native complex and pseudo complex inputs!

**Review by vincentqb — APPROVED** [review]
thx @anjali411 for the ping -- LGTM too! (with minor phrasing as you pointed out)

About the serialization test:
- @anjali411: are there tests on the pytorch-side around serialization? can/should we make the assumption the serialization is covered on that front? say by supporting the operations in [the doc](https://pytorch.org/docs/stable/notes/serialization.html)?
- Since adding a test for serialization could eventually catch errors especially when changing complex dtype parameters (say across versions), I can see this could help catch errors. However, since the parameters in [`TimeStretch`](https://github.com/pytorch/audio/blob/master/torchaudio/transforms.py#L727) are not changing for this pull request, I do not expect the `nn.Module` state to break BC compatibility with this pull request (say a user had saved the state prior to this pull request and now re-load it after this pull request).

File: torchaudio/transforms.py [hunk 5]
```diff
@@ -729,26 +729,24 @@ def __init__(self,
...
+                raise ValueError(
+                    "If no fixed_rate is specified, must pass a valid rate to the forward method.")
             rate = self.fixed_rate
-            if rate is None:
-                raise ValueError("If no fixed_rate is specified"
-                                 ", must pass a valid rate to the forward method.")
         else:
             rate = overriding_rate
-
-        if rate == 1.0:
```
Comment (anjali411) [inline torchaudio/transforms.py:749]: could you clarify this change?

**Comment by anjali411** [comment]
> thx @anjali411 for the ping -- LGTM too! (with minor phrasing as you pointed out)
> 
> About the serialization test:
> 
> * @anjali411: are there tests on the pytorch-side around serialization? can/should we make the assumption the serialization is covered on that front? say by supporting the operations in [the doc](https://pytorch.org/docs/stable/notes/serialization.html)?

No, because the JIT support for complex numbers is very new and some may say, it's still "under construction". 

> * Since adding a test for serialization could eventually catch errors especially when changing complex dtype parameters (say across versions), I can see this could help catch errors. However, since the parameters in [`TimeStretch`](https://github.com/pytorch/audio/blob/master/torchaudio/transforms.py#L727) are not changing for this pull request, I do not expect the `nn.Module` state to break BC compatibility with this pull request (say a user had saved the state prior to this pull request and now re-load it after this pull request).

That makes total sense and I agree. But I think it would be nice to have serialization tests, just in general, and especially for a sanity check now that we are migrating to complex types.

File: torchaudio/transforms.py (hunk 5 above)
Comment (mthrok) [reply torchaudio/transforms.py:749]: `TimeStretch` is a thin wrapper around `phase_vocoder`. and not altering the input Tensor when `rate == 1.0` is a valid branching for `phase_vocoder` itself, not just `TimeStretch`. so it makes more sense if it's in `phase_vocoder`.

**Comment by mthrok** [comment]
@anjali411 I am merging this so that I can resolve the conflict in other PRs. Let me know if you have further update, we can follow up on it in a separate PR.

File: test/torchaudio_unittest/functional/functional_impl.py [hunk 6]
```diff
@@ -89,3 +91,39 @@ def test_grad_at_zero(self, power):
...
+        phase_advance = torch.linspace(
+            0,
+            np.pi * hop_length,
+            num_freq,
+            dtype=self.real_dtype, device=self.device)[..., None]
+
+        spec_stretch = F.phase_vocoder(spec, rate=rate, phase_advance=phase_advance)
+
+        assert spec.dim() == spec_stretch.dim()
+        expected_shape = torch.Size([batch_size, num_freq, int(np.ceil(num_frames / rate))])
```
Comment (cpuhrsch) [inline test/torchaudio_unittest/functional/functional_impl.py:127]: FWIW torch also supports [ceil](https://pytorch.org/docs/stable/generated/torch.ceil.html)

**Pull request merged**